qutebrowser/browser/history.py

```diff
@@ -159,19 +168,33 @@ def __init__(self, progress, parent=None):
         self.completion = CompletionHistory(parent=self)
         self.metainfo = CompletionMetaInfo(parent=self)
 
-        if sql.Query('pragma user_version').run().value() < _USER_VERSION:
-            self.completion.delete_all()
+        rebuild_completion = False
+
+        if sql.db_user_version != sql.USER_VERSION:
+            # If the DB user version changed, run a full cleanup and rebuild the
+            # completion history.
+            #
+            # In the future, this could be improved to only be done when actually needed
+            # - but version changes happen very infrequently, rebuilding everything
+            # gives us less corner-cases to deal with, and we can run a VACUUM to make
+            # things smaller.
+            #
+            # If no history exists, we skip cleaning it up.
+            if self:
```

Comment:
Yep, Python [falls back on `__len__` if there is no `__bool__`](https://docs.python.org/3/reference/datamodel.html#object.__bool__):

> When [`__bool__`] method is not defined, `__len__()` is called, if it is defined, and the object is considered true if its result is nonzero. 

In 4d893e7 I added a `SqlQuery.__bool__` to make this more explicit (and also a tiny bit faster).

As for calling it in `__init__`, I suppose it'd be a bit cleaner to do `if not self: return` in `_cleanup_history()` instead, so it's a bit more self-contained (no pun intended). I'll do that once I get back to this, need to take care of something else right now.

Reply:
Turns out this actually isn't needed at all because of some other changes I did in the meantime... I removed it in 23847d5e0e36feeb6ecc4acd6e5e7500441a887a.